src/librustc/mir/tcx.rs

```diff
@@ -48,21 +45,18 @@ impl<'a, 'gcx, 'tcx> PlaceTy<'tcx> {
     /// Note that the resulting type has not been normalized.
     pub fn field_ty(self, tcx: TyCtxt<'a, 'gcx, 'tcx>, f: &Field) -> Ty<'tcx>
     {
-        // Pass `0` here so it can be used as a "default" variant_index in first arm below
-        let answer = match (self, VariantIdx::new(0)) {
-            (PlaceTy::Ty {
-                ty: &ty::TyS { sty: ty::TyKind::Adt(adt_def, substs), .. } }, variant_index) |
-            (PlaceTy::Downcast { adt_def, substs, variant_index }, _) => {
+        let answer = match self.to_ty().sty {
+            ty::TyKind::Adt(adt_def, substs) => {
+                let variant_index = match &self {
+                    PlaceTy::Ty { .. } => VariantIdx::new(0),
+                    PlaceTy::Downcast { ty: _, variant_index } => *variant_index,
+                };
```

Comment:
This is subtly incomplete, but pre-existing.
The non-downcast case should use `.non_enum_variant()` to get the variant, while downcasts should assert the ADT is an enum.

Reply:
done